Add option to set GeoPDF group name for items

nyalldawson opened the pull request:
This new setting, located in the layout item "Rendering" section, allows users to set an optional "group name" for use in GeoPDF
exports. When set, a matching layer tree group will be created in the exported GeoPDF and the item will only be visible when
this group is checked.

This allows content to be selectively displayed as a group by viewers of the GeoPDF. Eg, it can allow extra layout content
such as descriptive labels or legends to only be shown when layers from the group are visible, making GeoPDF export much
more flexible.

![image](https://github.com/qgis/QGIS/assets/1829991/6a2239af-5c9a-4dc8-881b-29871e6d363e)


Sponsored by Rubicon Concierge Real Estate Services

nyalldawson committed:
[feature][layouts] Add option to set geopdf group name for items

This new setting, located in the layout item "Rendering" section,
allows users to set an optional "group name" for use in GeoPDF
exports. When set, a matching layer tree group will be created
in the exported GeoPDF and the item will only be visible when
this group is checked.

This allows content to be selectively displayed as a group by
viewers of the GeoPDF. Eg, it can allow extra layout content
such as descriptive labels or legends to only be shown when
layers from the group are visible, making GeoPDF export much
more flexible.

nyalldawson committed:
Fix handling of mutually exclusive groups in GeoPDF export

When custom groups are present mutually exclusive group handling
was not working correctly - it was incorrectly assuming every
group must be a mutually exclusive group.

I've refactored the whole composition XML generation to be less
fragile and more future proof, and to ensure cleaner layer trees
are written to the XML.

Also added many additional tests.

nyalldawson committed:
Spelling

nirvn commented on src/ui/layout/qgslayoutitemwidgetbase.ui:
```diff
@@ -728,27 +717,52 @@
+      <item row="4" column="0">
+       <widget class="QLabel" name="label_5">
+        <property name="text">
+         <string>GeoPDF group</string>
```
@nyalldawson , could we try to avoid being GeoPDF specific for this label (even though ATM it's every only used for GeoPDF)?

nyalldawson replied:
@nirvn I've just removed the generic code now -- when I tried to make this generic it just ended up messy, as the requirements for eg SVG groups are very different to GeoPDF groups

nirvn replied:
@nyalldawson , K, then GeoPDF-specific label there makes more sense.

nyalldawson committed:
Drop generic option, go with custom property specific for GeoPDF exports

It's too messy to try to handle this in a format-agnostic way, as
the requirement for GeoPDF is very unique to GeoPDF (ie SVG and other
formats don't export as mini-read-only-projects with custom layer
trees like GeoPDF does)

nyalldawson committed:
Remove broken see also link
